Fix setting min/max decimal values for single band gray renderer

rnousia opened the pull request:
Fixes issue #50861

## Description

The issue was with layer styling dock updating QLineEdit's contents too fast because raster renderer is called at some point when raster min/max values are changed. Min/max decimal values are being converted from string to text in a way that truncates decimal point before decimal digits have been entered, thus making it difficult for the user to enter decimal values. Starting to type "2." will change the value automatically to "2" unless user is typing very fast.

The same issue is found at least in QgsMultiBandColorRendererWidget min/max values (tested with QGIS) and possibly in QgsPointCloudRgbRendererWidget (didn't test this but code looks similar). If the solution is ok, I could fix them (or at least QgsMultiBandColorRendererWidget) in this PR also.

The issue happens when QLineEdit's text change triggers rendering and the values are at some point set back to QLineEdit text (input converted from string -> to double -> back to string).


Renderer converts the string input to double:
```c++
// src/gui/raster/qgssinglebandgrayrendererwidget.cpp
QgsRasterRenderer *QgsSingleBandGrayRendererWidget::renderer()
    //...

    // Here "2." gets converted to 2 (QgsDoubleValidator handles it)
    QgsContrastEnhancement *e = new QgsContrastEnhancement( ( Qgis::DataType )(
        provider->dataType( band ) ) );
    e->setMinimumValue( QgsDoubleValidator::toDouble( mMinLineEdit->text() ) );
    e->setMaximumValue( QgsDoubleValidator::toDouble( mMaxLineEdit->text() ) );
    
    ///...
```

refreshAfterStyleChanged is triggered at some point during editing values in layer styling dock:
```c++
// src/gui/raster/qgssinglebandgrayrendererwidget.cpp
void QgsRendererRasterPropertiesWidget::refreshAfterStyleChanged()
{
  if ( mRendererWidget )
  {
    // Here renderer is called -> input of "2." sets QgsContrastEnhancement.minimumValue to 2
    QgsRasterRenderer *renderer = mRasterLayer->renderer();
    
    // ..
    
    else if ( QgsSingleBandGrayRenderer *sbgr = dynamic_cast<QgsSingleBandGrayRenderer *>( renderer ) )
    {
      const QgsContrastEnhancement *ce = sbgr->contrastEnhancement();
      if ( ce )
      {
        // Here values are set back to QLineEdit (see snippet below)
        mRendererWidget->setMin( QLocale().toString( ce->minimumValue() ) );
        mRendererWidget->setMax( QLocale().toString( ce->maximumValue() ) );
      }
    }
  }
}
```

```c++
// src/gui/raster/qgssinglebandgrayrendererwidget.cpp
void QgsSingleBandGrayRendererWidget::setMin( const QString &value, int )
{
  mDisableMinMaxWidgetRefresh = true;
  // Here value is set back to QLineEdit widget: 2 becomes "2" (which is 
  // seen in layer styling dock as decimal point disappearing while typing)
  mMinLineEdit->setText( value );
  mDisableMinMaxWidgetRefresh = false;
}
```

I was looking into this issue with @jgrocha  who came up the solution to check if `QgsDoubleValidator::toDouble` changes the input during conversion and only trigger rendering if conversion does not change the input. This allows user to type decimal point without re-rendering the raster (re-rendering happens when first digit is typed).

I was trying to find tests related to the raster renderer widgets but did not find one for QgsSingleBandGrayRendererWidget. I am unsure whether this needs an unit test as it is GUI related. If it does need a test, where should the test be added?

rnousia committed:
Fix setting min/max decimal values for single band gray renderer

Fixes issue #50861

jgrocha commented:
Thank you @rnousia for this PR.

I've checkout it on my side and it fixes the issue for the single band renderer. :partying_face:

As you pointed out, this same limitations happens in the multiband renderer.  I've tested with a point cloud layer too. There is no problem writing values with decimals in the RGB renderer (min/max values).

So, while you are working on this, I would suggest to add another commit to address the multiband renderer issue too.

rnousia committed:
Fix setting decimal values to multiband color renderer

Add new static member function to QgsDoubleValidator to check if string to double
conversion changes the input string.

rnousia committed:
Add SIP_SKIP
